graphix/graphsim/rxgraphstate.py

```diff
@@ -0,0 +1,144 @@
+        return iter(ret)
+
+    def neighbors(self, node) -> iter:
+        nidx = self.nodes.get_node_index(node)
+        return iter(self._graph.neighbors(nidx))
+
+    def subgraph(self, nodes: list) -> rx.PyGraph:
+        nidx = [self.nodes.get_node_index(n) for n in nodes]
+        return self._graph.subgraph(nidx)
+
+    def number_of_edges(self, u: int | None = None, v: int | None = None) -> int:
+        if u is None and v is None:
+            return len(self.edges)
+        elif u is None or v is None:
+            raise ValueError("u and v must be specified together")
+        uidx = self.nodes.get_node_index(u)
+        vidx = self.nodes.get_node_index(v)
+        return len(self._graph.get_all_edge_data(uidx, vidx))
+
+    def adjacency(self) -> iter:
+        ret = []
+        for n in self.nodes:
+            nidx = self.nodes.get_node_index(n)
+            adjacency_dict = self._graph.adj(nidx)
+            new_adjacency_dict = {}
+            for nidx, _ in adjacency_dict.items():
+                new_adjacency_dict[self.nodes.get_node_index(nidx)] = {}  # replace None with {}
+            ret.append((n, new_adjacency_dict))
+        return iter(ret)
+
+    def remove_node(self, node: int) -> None:
+        nidx = self.nodes.get_node_index(node)
+        self._graph.remove_node(nidx)
+        self.nodes.remove_node(node)
+        edge_list = list(self.edges)
+        for e in edge_list:
+            if node in e:
+                self.edges.remove_edge(e)
+
+    def remove_nodes_from(self, nodes: list[int]) -> None:
+        for n in nodes:
+            self.remove_node(n)
+
+    def remove_edge(self, u: int, v: int) -> None:
+        uidx = self.nodes.get_node_index(u)
+        vidx = self.nodes.get_node_index(v)
+        self._graph.remove_edge(uidx, vidx)
+        self.edges.remove_edge((u, v))
+
+    def remove_edges_from(self, edges: list[tuple[int, int]]) -> None:
+        for e in edges:
+            self.remove_edge(e[0], e[1])
+
+    def add_nodes_from(self, nodes: list[int]):
+        node_indices = self._graph.add_nodes_from([(n, {"loop": False, "sign": False, "hollow": False}) for n in nodes])
+        for nidx in node_indices:
+            self.nodes.add_node(self._graph[nidx][0], self._graph[nidx][1], nidx)
+
+    def add_edges_from(self, edges):
+        for u, v in edges:
+            # adding edges may add new nodes
+            if u not in self.nodes:
+                nidx = self._graph.add_node((u, {"loop": False, "sign": False, "hollow": False}))
+                self.nodes.add_node(self._graph[nidx][0], self._graph[nidx][1], nidx)
+            if v not in self.nodes:
+                nidx = self._graph.add_node((v, {"loop": False, "sign": False, "hollow": False}))
+                self.nodes.add_node(self._graph[nidx][0], self._graph[nidx][1], nidx)
+            uidx = self.nodes.get_node_index(u)
+            vidx = self.nodes.get_node_index(v)
+            eidx = self._graph.add_edge(uidx, vidx, None)
```

Comment:
This part seems to be the bottleneck of the rustworkx-based simulator. I'm sorry for not having any suggestions, but could you consider if it can be improved?

Reply:
I think I figured out the problem.
It seems like I should have defined `__contains__` method for `NodeList` and `EdgeList`, otherwise the `in` uses `__iter__` method to make `in` work and the conversion of `set` to `iter` affected the performance.

```python
class TestIn1:
    def __init__(self, x):
        self.x = set(x)

    def __iter__(self):
        return iter(self.x)

ins1 = TestIn1(list(range(10000000)))
```

```python
%%timeit
59999 in ins1
```
out: `406 µs ± 4.04 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)`

```python
class TestIn2:
    def __init__(self, x):
        self.x = set(x)

    def __iter__(self):
        return iter(self.x)

    def __contains__(self, item):
        return item in self.x

ins2 = TestIn2(list(range(10000000)))
```

```python
%%timeit
59999 in ins2
```
out:`45.8 ns ± 0.414 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)`